File: Sem6/JPP/Lista4/CPP/zad1.cpp
#include <iostream>
#include <thread>
#include <mutex>
#include <vector>
#include <chrono>
#include <iomanip>
#include <sstream>
#include <atomic>
#include <random>

using namespace std;
using namespace chrono;

class Logger
{
private:
    mutex mtx_log;
    steady_clock::time_point startTime;

    string getTimeStamp()
    {
        auto now = steady_clock::now();
        auto elapsed = duration_cast<milliseconds>(now - startTime);
        stringstream ss;
        ss << "[" << setw(6) << elapsed.count() << "ms]";
        return ss.str();
    }

public:
    Logger() : startTime(steady_clock::now()) {};

    void log(int philosopherId, const string &action)
    {
        lock_guard<mutex> lock(mtx_log);
        cout << getTimeStamp() << " Filozof" << philosopherId << ": " << action << "\n";
    }
};

class Fork
{
private:
    mutex mtx;
    bool isClean;
    int ownerId;

public:
    Fork() : isClean(true), ownerId(-1) {}
    bool isForkClean() const { return isClean; }
    void setDirty() { isClean = false ;}
    void setClean() { isClean = true; }

    bool tryTake(int philosopherId, bool &wasDirty)
    {
        if (mtx.try_lock())
        {
            wasDirty = !isClean;
            ownerId = philosopherId;
            return true;
        }
        return false;
    }
    void take(int philosopherId, bool &wasDirty)
    {
        mtx.lock();
        wasDirty = !isClean;
        ownerId = philosopherId;
    }
    void put(int philosopherId)
    {
        isClean = false;
        ownerId = -1;
        mtx.unlock();
    }
};

class Philosopher
{
private:
    int id;
    int mealsToEat;
    atomic<int> &failedMeals;
    vector<Fork> &forks;
    Logger &logger;
    random_device rd;
    mt19937 gen;
    uniform_int_distribution<> thinkDist;
    uniform_int_distribution<> eatDist;

public:
    void think()
    {
        logger.log(id, "Myśli...");
        this_thread::sleep_for(milliseconds(thinkDist(gen)));
    }
    void eat()
    {
        logger.log(id, "JE OBIAD! (numer " + to_string(mealsToEat - failedMeals.load()) + ")");
        this_thread::sleep_for(milliseconds(eatDist(gen)));
    }

    bool takeForks()
    {
        int firstFork = id;
        int secondFork = (id + 1) % forks.size();

        if (firstFork > secondFork)
        {
            swap(firstFork, secondFork);
        }

        bool wasDirty1, wasDirty2;

        forks[firstFork].take(id, wasDirty1);
        auto start = steady_clock::now();
        while (duration_cast<milliseconds>(steady_clock::now() - start).count() < 100)
        {
            if (forks[secondFork].tryTake(id, wasDirty2))
            {
                if (wasDirty1)
                    forks[firstFork].setClean();
                if (wasDirty2)
                    forks[secondFork].setClean();

                logger.log(id, "Udało się wziąć widelce " + to_string(firstFork) +
                                   " i " + to_string(secondFork));
                return true;
            }
            this_thread::sleep_for(milliseconds(10));
        }

        forks[firstFork].put(id);
        logger.log(id, "Nie udało się wziąć widelców");
        return false;
    }

public:
    Philosopher(int i, int meals, atomic<int> &failed, vector<Fork> &f, Logger &l) : id(i), mealsToEat(meals), failedMeals(failed), forks(f), logger(l), gen(rd()), thinkDist(50, 200), eatDist(100, 300)
    {
        failedMeals = 0;
    }
    void dine()
    {
        int mealsEaten = 0;
        while (mealsEaten < mealsToEat)
        {
            think();
            if (takeForks())
            {
                eat();
                int firstFork = id;
                int secondFork = (id + 1) % forks.size();

                forks[secondFork].put(id);
                forks[firstFork].put(id);
                mealsEaten++;

                logger.log(id, "Skończył posiłek " + to_string(mealsEaten) + "/" + to_string(mealsToEat));
            }
            else
            {
                failedMeals++;
                logger.log(id, "GŁODUJE! (nie udało się " + to_string(failedMeals.load()) + " razy)");
            }
        }
        logger.log(id, "Skończył wszystkie posiłki");
    }
};

int main(int argc, char *argv[])
{
    int numOfPhilosophers = 5;
    int mealsPerPhilosopher = 5;

    if (argc == 3)
    {
        numOfPhilosophers = stoi(argv[1]);
        mealsPerPhilosopher = stoi(argv[2]);
    }
    else
    {
        cout << "./zad1 <numOfPhilosophers> <mealsPerPhilosopher> \n";
        return 0;
    }

    cout << "Liczba filozofów " << numOfPhilosophers << "\n";
    cout << "Liczba posiłków na filozofa " << mealsPerPhilosopher << "\n";

    Logger logger;
    vector<thread> philosophers;
    vector<atomic<int>> failedMeals(numOfPhilosophers);
    vector<Fork> forks(numOfPhilosophers);
    for (int i = 0; i < numOfPhilosophers; i++)
    {
        philosophers.emplace_back([i, mealsPerPhilosopher, &failedMeals, &forks, &logger]()
                                  {
                Philosopher p(i, mealsPerPhilosopher, failedMeals[i], forks, logger);
                p.dine(); });
    }
    for (auto& p : philosophers) {
        p.join();
    }
 // Podsumowanie
    cout << "\n=====================================" << endl;
    cout << "PODSUMOWANIE:" << endl;
    int totalFailed = 0;
    for (int i = 0; i < numOfPhilosophers; i++) {
        cout << "Filozof " << i << ": nie udało się zjeść " << failedMeals[i].load() << " razy" << endl;
        totalFailed += failedMeals[i].load();
    }
    
    cout << "\nCzy system był sprawiedliwy? ";
    int maxFailed = 0, minFailed = mealsPerPhilosopher;
    for (int i = 0; i < numOfPhilosophers; i++) {
        maxFailed = max(maxFailed, failedMeals[i].load());
        minFailed = min(minFailed, failedMeals[i].load());
    }
    
    if (maxFailed - minFailed <= 1) {
        cout << "TAK - wszystkie filozofy głodowały podobnie często" << endl;
    } else {
        cout << "NIE - filozofowie głodowali z różną częstotliwością (różnica " 
             << (maxFailed - minFailed) << ")" << endl;
    }
    
    cout << "Łączna liczba nieudanych prób: " << totalFailed << endl;
    cout << "=====================================" << endl;
    
    return 0;
}

File: Sem6/JPP/Lista4/GO/zad1.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"sync"
	"sync/atomic"
	"time"
)

type Logger struct {
	mu        sync.Mutex
	startTime time.Time
}

func NewLogger() *Logger {
	return &Logger{
		startTime: time.Now(),
	}
}

func (l *Logger) getTimestamp() string {
	elapsed := time.Since(l.startTime).Milliseconds()
	return fmt.Sprintf("[%6dms]", elapsed)
}

func (l *Logger) Log(philosopherId int, action string) {
	l.mu.Lock()
	defer l.mu.Unlock()
	fmt.Printf("%s Filozof%d: %s\n", l.getTimestamp(), philosopherId, action)
}

type Fork struct {
	mu              sync.Mutex
	isClean         bool
	ownerId         int
	waitingPriority int32 
}

func NewFork() *Fork {
	return &Fork{
		isClean:         true,
		ownerId:         -1,
		waitingPriority: -1,
	}
}

func (f *Fork) IsClean() bool {
	return f.isClean
}

func (f *Fork) SetDirty() {
	f.isClean = false
}

func (f *Fork) SetClean() {
	f.isClean = true
}

func (f *Fork) TryTake(philosopherId int, wasDirty *bool) bool {
	if f.mu.TryLock() {
		*wasDirty = !f.isClean
		f.ownerId = philosopherId
		f.waitingPriority = -1
		return true
	}
	return false
}
func (f *Fork) TryTakeWithPriority(philosopherId int, hungerLevel int32, wasDirty *bool) bool {
	if atomic.LoadInt32(&f.waitingPriority) > hungerLevel {
		return false
	}

	if f.mu.TryLock() {
		*wasDirty = !f.isClean
		f.ownerId = philosopherId
		atomic.StoreInt32(&f.waitingPriority, -1)
		return true
	}

	current := atomic.LoadInt32(&f.waitingPriority)
	if hungerLevel > current {
		atomic.CompareAndSwapInt32(&f.waitingPriority, current, hungerLevel)
	}
	return false
}

// Take - blokujące branie widelca
func (f *Fork) Take(philosopherId int, wasDirty *bool) {
	f.mu.Lock()
	*wasDirty = !f.isClean
	f.ownerId = philosopherId
	atomic.StoreInt32(&f.waitingPriority, -1)
}

// Put - zwolnienie widelca
func (f *Fork) Put(philosopherId int) {
	f.isClean = false
	f.ownerId = -1
	atomic.StoreInt32(&f.waitingPriority, -1)
	f.mu.Unlock()
}

// Philosopher - filozof
type Philosopher struct {
	id                   int
	mealsToEat           int32
	failedMeals          *int32
	forks                []*Fork
	logger               *Logger
	thinkDist            [2]int
	eatDist              [2]int
	consecutiveFailures  int32
	totalFailedAttempts  int32
}

func NewPhilosopher(id int, mealsToEat int32, failedMeals *int32, forks []*Fork, logger *Logger) *Philosopher {
	return &Philosopher{
		id:                  id,
		mealsToEat:          mealsToEat,
		failedMeals:         failedMeals,
		forks:               forks,
		logger:              logger,
		thinkDist:           [2]int{50, 200},
		eatDist:             [2]int{100, 300},
		consecutiveFailures: 0,
		totalFailedAttempts: 0,
	}
}

func (p *Philosopher) getHungerLevel() int32 {
	level := p.consecutiveFailures * 10
	if level > 100 {
		level = 100
	}
	return level
}

func (p *Philosopher) think() {
	baseTime := p.thinkDist[0] + rand.Intn(p.thinkDist[1]-p.thinkDist[0])
	thinkTime := baseTime

	if p.consecutiveFailures > 3 {
		thinkTime = baseTime / 2
		p.logger.Log(p.id, fmt.Sprintf("Myśli... (GŁODNY! czas skrócony do %dms)", thinkTime))
	} else if p.consecutiveFailures > 0 {
		thinkTime = baseTime * 7 / 10
		p.logger.Log(p.id, fmt.Sprintf("Myśli... (lekko głodny, czas %dms)", thinkTime))
	} else {
		p.logger.Log(p.id, fmt.Sprintf("Myśli... (spokojnie, czas %dms)", thinkTime))
	}

	time.Sleep(time.Duration(thinkTime) * time.Millisecond)
}

func (p *Philosopher) eat(mealsEaten int32) {
	p.logger.Log(p.id, fmt.Sprintf("JE OBIAD! (posiłek %d/%d)", mealsEaten+1, p.mealsToEat))
	eatTime := p.eatDist[0] + rand.Intn(p.eatDist[1]-p.eatDist[0])
	time.Sleep(time.Duration(eatTime) * time.Millisecond)
}

func (p *Philosopher) takeForks() bool {
	N := len(p.forks)
	leftFork := p.id
	rightFork := (p.id + 1) % N

	firstFork := min(leftFork, rightFork)
	secondFork := max(leftFork, rightFork)

	var wasDirty1, wasDirty2 bool
	hungerLevel := p.getHungerLevel()

	p.forks[firstFork].Take(p.id, &wasDirty1)

	timeout := 100 + (int(p.consecutiveFailures) * 20)
	if timeout > 500 {
		timeout = 500
	}

	if p.consecutiveFailures > 5 {
		p.logger.Log(p.id, fmt.Sprintf("OSTRZEŻENIE: Bardzo głodny! timeout=%dms", timeout))
	}

	start := time.Now()
	attempts := 0

	for time.Since(start).Milliseconds() < int64(timeout) {
		var success bool

		if p.consecutiveFailures > 2 {
			success = p.forks[secondFork].TryTakeWithPriority(p.id, hungerLevel, &wasDirty2)
		} else {
			success = p.forks[secondFork].TryTake(p.id, &wasDirty2)
		}

		if success {
			if wasDirty1 {
				p.forks[firstFork].SetClean()
			}
			if wasDirty2 {
				p.forks[secondFork].SetClean()
			}

			atomic.StoreInt32(&p.consecutiveFailures, 0)

			p.logger.Log(p.id, fmt.Sprintf("Udało się wziąć widelce %d i %d (po %d próbach)", firstFork, secondFork, attempts))
			return true
		}

		attempts++
		if attempts%5 == 0 && p.consecutiveFailures > 2 {
			p.logger.Log(p.id, fmt.Sprintf("Ciągle czekam na widelec %d... (próba %d)", secondFork, attempts))
		}

		time.Sleep(10 * time.Millisecond)
	}

	p.forks[firstFork].Put(p.id)
	atomic.AddInt32(&p.consecutiveFailures, 1)
	atomic.AddInt32(&p.totalFailedAttempts, 1)
	atomic.AddInt32(p.failedMeals, 1)

	p.logger.Log(p.id, fmt.Sprintf("Nie udało się wziąć widelców (głód poziom %d, timeout %dms)", 
		atomic.LoadInt32(&p.consecutiveFailures), timeout))
	return false
}

func (p *Philosopher) Dine() {
	var mealsEaten int32 = 0

	for mealsEaten < p.mealsToEat {
		p.think()

		if p.takeForks() {
			p.eat(mealsEaten)

			N := len(p.forks)
			leftFork := p.id
			rightFork := (p.id + 1) % N

			p.forks[rightFork].Put(p.id)
			p.forks[leftFork].Put(p.id)
			mealsEaten++

			p.logger.Log(p.id, fmt.Sprintf("Skończył posiłek %d/%d (łącznie głodował %d razy)", 
				mealsEaten, p.mealsToEat, atomic.LoadInt32(&p.totalFailedAttempts)))
		} else {
			p.logger.Log(p.id, fmt.Sprintf("GŁODUJE! (nie udało się %d razy)", atomic.LoadInt32(p.failedMeals)))
		}
	}

	p.logger.Log(p.id, fmt.Sprintf("ZAKOŃCZYŁ WSZYSTKIE POSIŁKI! (głodował %d razy)", 
		atomic.LoadInt32(&p.totalFailedAttempts)))
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	numOfPhilosophers := 5
	mealsPerPhilosopher := 5

	if len(os.Args) == 3 {
		fmt.Sscan(os.Args[1], &numOfPhilosophers)
		fmt.Sscan(os.Args[2], &mealsPerPhilosopher)
	} else {
		fmt.Println("./zad1 <numOfPhilosophers> <mealsPerPhilosopher>")
		fmt.Println("Przykład: ./zad1 7 3")
		return
	}


	rand.Seed(time.Now().UnixNano())

	fmt.Printf("\n=== PROBLEM UCZTUJĄCYCH FILOZOFÓW ===\n")
	fmt.Printf("Liczba filozofów: %d\n", numOfPhilosophers)
	fmt.Printf("Posiłków na filozofa: %d\n", mealsPerPhilosopher)
	fmt.Printf("Mechanizm: czyste/brudne widelce + priorytet dla głodujących\n")
	fmt.Printf("======================================\n\n")

	logger := NewLogger()
	var wg sync.WaitGroup

	failedMeals := make([]int32, numOfPhilosophers)
	forks := make([]*Fork, numOfPhilosophers)
	for i := 0; i < numOfPhilosophers; i++ {
		forks[i] = NewFork()
	}

	for i := 0; i < numOfPhilosophers; i++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			philosopher := NewPhilosopher(id, int32(mealsPerPhilosopher), &failedMeals[id], forks, logger)
			philosopher.Dine()
		}(i)
	}

	wg.Wait()

	fmt.Printf("\n=========================================\n")
	fmt.Printf("PODSUMOWANIE:\n")
	totalFailed := int32(0)
	for i := 0; i < numOfPhilosophers; i++ {
		fmt.Printf("Filozof %d: nie udało się zjeść %d razy\n", i, failedMeals[i])
		totalFailed += failedMeals[i]
	}

	fmt.Printf("\nCzy system był sprawiedliwy? ")
	maxFailed := int32(0)
	minFailed := int32(mealsPerPhilosopher)
	for i := 0; i < numOfPhilosophers; i++ {
		if failedMeals[i] > maxFailed {
			maxFailed = failedMeals[i]
		}
		if failedMeals[i] < minFailed {
			minFailed = failedMeals[i]
		}
	}

	if maxFailed-minFailed <= 1 {
		fmt.Printf("TAK  - wszyscy filozofowie głodowali podobnie często\n")
	} else {
		fmt.Printf("NIE  - filozofowie głodowali z różną częstotliwością (różnica %d)\n", maxFailed-minFailed)
	}

	fmt.Printf("Łączna liczba nieudanych prób: %d\n", totalFailed)
	fmt.Printf("Średnia nieudanych prób na filozofa: %d\n", totalFailed/int32(numOfPhilosophers))
	fmt.Printf("=========================================\n")
}